src/class-main.php

```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
       return;
     }
 
-    $classes = $this->get_service_classes();
+    $this->services = $this->get_service_classes_with_di();
 
-    $this->services = array_map(
-      [ $this, 'instantiate_service' ],
-      $classes
-    );
     array_walk(
       $this->services,
-      function( Service $service ) {
-        $service->register();
+      function( $class ) {
+        if ( ! $class instanceof Service ) {
+          return;
+        }
+
+        $class->register();
+      }
+    );
+  }
+
+  /**
+   * Return array of services with Dependency Injection parameters.
+   *
+   * @return array
+   *
+   * @since 0.7.0 Init
+   */
+  private function get_service_classes_with_di() : array {
+    $services = $this->get_service_classes_prepared_array();
+
+    $container = $this->get_di_container( $services );
+
+    return array_map(
+      function( $class ) use ( $container ) {
+        return $container->get( $class );
+      },
+      array_keys( $services )
+    );
+  }
+
+  /**
+   * Get services classes array ad prepare it.
+   * Key should be class name, and value empty array or DI items.
+   *
+   * @return array
+   *
+   * @since 0.7.0 Init.
+   */
+  private function get_service_classes_prepared_array() : array {
+    $output = [];
+    foreach ( $this->get_service_classes() as $key => $value ) {
```

Comment:
```suggestion
    foreach ( $this->get_service_classes() as $class => $dependencies ) {
```

Makes it more clear what you are looping through.

Reply:
👍